### PR TITLE
Adding light position configurability for `surface` and `mesh3d`, and general lighting configurability for the latter

### PR DESCRIPTION
**Pull request by monfera**
Part of the fix to https://github.com/plotly/plotly.js/issues/552

The fix isn't actually to the 'plotly.js' repo, but rather, the dependencies are changed such that epsilon values can be passed on to them. The alternative fix in the dependencies (changing the epsilon in the `normals` repo) was deemed less good because the epsilon is better be configurable in the dependencies.

PRs in dependent issues:
https://github.com/mikolalysenko/normals/pull/2
https://github.com/gl-vis/gl-mesh3d/pull/5

The other part of the fix is even less to do with `plotly.js` (except bumping the dependency version number): https://github.com/gl-vis/gl-mesh3d/pull/5/commits/b901f896ee18cb9fc79dc129a8a6d8225a5a199a

Adding epsilon configuration in order to pass it on to newly modified dependencies that now take these new params. The default values are set such that 3D mesh models with minute triangles are also properly included in the GL specular light reflection calculation.

WIP because:
- it assumes the PR acceptance, merge in and npm update of a dependency; which in turn assumes a similar change to its own dependency
- attribute and image test cases are not yet added
- anything I forgot?

### CODE REVIEW

**Comment by monfera** [comment]
As of now, the CI check obviously fails for the reason that the package dependencies haven't yet been bumped and pushed to the NPM repo so it doesn't find `gl-mesh3d v1.1.0`.

**Comment by monfera** [comment]
Adding @mikolalysenko @etpinard @bpostlethwaite

File: src/traces/surface/attributes.js [hunk 1]
```diff
@@ -182,6 +206,20 @@ module.exports = {
             min: 0.00,
             max: 5.00,
             dflt: 0.2
+        },
+        vertexnormalsepsilon: {
```
Comment (etpinard) [inline src/traces/surface/attributes.js]: maybe `vertexnormal` and `facenormal` would suffice?

File: src/traces/surface/defaults.js [hunk 2]
```diff
@@ -50,13 +50,19 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
...
+    [
+        'lighting.ambient',
+        'lighting.diffuse',
+        'lighting.specular',
+        'lighting.roughness',
+        'lighting.fresnel',
+        'lightposition.x',
+        'lightposition.y',
+        'lightposition.z',
+        'hidesurface',
```
Comment (etpinard) [inline src/traces/surface/defaults.js]: looks like you're not coercing `vertexnormalsepsilon` and `facenormalsepsilon` in surface traces, is that on purpose?

File: src/traces/surface/attributes.js [hunk 3]
```diff
@@ -147,6 +147,30 @@ module.exports = {
         ].join(' ')
     },
 
+    lightposition: {
+        x: {
+            valType: 'number',
+            role: 'style',
+            min: -1e5,
+            max: 1e5,
+            dflt: 1e5
```
Comment (etpinard) [inline src/traces/surface/attributes.js]: we'll need to add `description` fields to all new attributes.

**Comment by etpinard** [comment]
@monfera looks like your patch led to image test diffs:

image

Is that on purpose?

**Comment by etpinard** [comment]
@monfera great work :tada: 

Thanks for bringing our `mesh3d` trace type up to its full potential. 

Before merging this in:

- [ ] agree on vertex normal and face normal epsilon attribute names
- [ ] update the baseline images that generated diffs
- [ ] add a image test mock showing off custom `lightposition`, and vertex/face normal epsilions. 
- [ ] add `description` fields to all added attributes

File: src/traces/surface/attributes.js (hunk 1 above)
Comment (monfera) [inline src/traces/surface/attributes.js]: @etpinard it's quite specific to to the epsilon error terms in the normals calculation and these names would be suggestive of passing on the normals directly.

File: src/traces/surface/attributes.js (hunk 3 above)
Comment (monfera) [inline src/traces/surface/attributes.js]: Thank you, yes, I'll add them!

File: src/traces/surface/defaults.js (hunk 2 above)
Comment (monfera) [inline src/traces/surface/defaults.js]: Correct, initially I wasn't going to shoot for covering the `surface3d` case, just `mesh3d`; the reason for adding the properties there is that `mesh3d` piggybacks on it and we may at some point need this feature for the surface anyway. I'd suggest let's keep this PR for the `mesh3d` as it's being expected today and I'd do a subsequent small PR for the `surface3d` case.

**Comment by monfera** [comment]
@etpinard yes, I knew about the image diff but haven't yet pushed the commit for its solution. Also I agree with your list and push commits, and pls. see my reply on the naming.

File: src/traces/surface/attributes.js (hunk 1 above)
Comment (etpinard) [inline src/traces/surface/attributes.js]: OK. Agreed. I won't be the most used attribute nor is it the most obvious ; might as well be verbose. :+1:

File: src/traces/surface/defaults.js (hunk 2 above)
Comment (etpinard) [inline src/traces/surface/defaults.js]: Adding the `vertexnormalsepsilon` and `facenormalsepsilon` functionality to mesh3d traces only is fine. :+1: 

But in this case, we should **not** add `vertexnormalsepsilon` and `facenormalsepsilon` to the  `surface` attributes to not confuse users reading attribute reference [schema](https://plot.ly/javascript/reference/).

To do so:

- remove `vertexnormalsepsilon` and `facenormalsepsilon` from the `surface` [attributes](https://github.com/plotly/plotly.js/pull/556/files#diff-b4784b514eb9a0e0ada46bac7f483a64R210)
- rewrite this [line](https://github.com/plotly/plotly.js/pull/556/files#diff-a9cf6c3c9e5f641b36f80ac4f8cb2dc7R128) as

```js
lighting: extendFlat({}, surfaceAtts.lighting, {
  vertexnormalepsilon: {
    /* */
  }, 
  facenormalepsilon: {
    /* */
  }
})
```

File: src/traces/surface/attributes.js [hunk 4]
```diff
@@ -177,49 +180,59 @@ module.exports = {
...
+            description: 'Represents the level that incident rays are reflected in a single direction, causing shine.'
         },
         roughness: {
             valType: 'number',
             role: 'style',
             min: 0.00,
             max: 1.00,
-            dflt: 0.5
+            dflt: 0.5,
+            description: 'Alters specular reflection; the rougher the surface, the wider and less contrasty the shine.'
```
Comment (etpinard) [inline src/traces/surface/attributes.js]: Wow. These descriptions are great. Thanks!

Comment (etpinard) [inline src/traces/surface/attributes.js]: cc @cldougl @chriddyp

**Comment by etpinard** [comment]
Yes yes. The new Snowden image is incredible:

image

File: src/traces/surface/attributes.js (hunk 4 above)
Comment (chriddyp) [inline src/traces/surface/attributes.js]: 🙇

**Comment by monfera** [comment]
The included test doesn't yet cover specular reflection yet. Snowden head shows the effects of the diffuse and ambient parameters, and the mesh color parameter is also exercised.
image

**Comment by monfera** [comment]
@etpinard regarding the coercion of `vertexnormalsepsilon` and  `facenormalsepsilon`: these turn out not to be used by the underlying `gl-surface3d` so I'm making it specific to `mesh3d`.

However I'm making commits to this PR and to `gl-surface3d` such that the directional light position is being used, see an upcoming `png` diff.

**Comment by monfera** [comment]
Snowden in a different lighting and orientation:
image
image

**Comment by monfera** [comment]
@etpinard I didn't want to fiddle with your PR checklist above but I think they were covered by late yesterday, and today's commits made light positions work with `surface` too. The epsilon values weren't needed as it uses a different specular calculation method.

File: src/traces/mesh3d/attributes.js [hunk 5]
```diff
@@ -125,7 +125,24 @@ module.exports = {
...
+        },
+        facenormalsepsilon: {
+            valType: 'number',
+            role: 'style',
+            min: 0.00,
+            max: 1,
+            dflt: 1e-6, // even the brain model doesn't appear to need finer than this
+            description: 'Epsilon for face normals calculation avoids math issues arising from degenerate geometry.'
+        }
+    }, surfaceAtts.lighting),
```
Comment (etpinard) [inline src/traces/mesh3d/attributes.js]: thanks!

**Comment by etpinard** [comment]
@monfera 

> The different lighting cast on the yield curve highlights an issue with the specific test case.

Can you elaborate? If I understand correctly, `lightning` on surface trace isn't perfect yet?

**Comment by etpinard** [comment]
@monfera Fantastic work. 

You win the PR of the week :trophy:

**Pull request merged**